File: src/backend/src/utils/validation.utils.ts
/**
 * Human Tasks:
 * 1. Review validation thresholds with security team
 * 2. Confirm image file size limits with infrastructure team
 * 3. Verify phone number validation patterns with international requirements
 * 4. Review password complexity requirements with security policy
 * 5. Validate coordinate precision requirements with location team
 */

// @ts-ignore validator@13.7.0
import validator from 'validator';
// @ts-ignore xss@1.0.11
import xss from 'xss';
import { ERROR_CODES } from '../constants/error-codes';
import { ERROR_MESSAGES } from '../constants/messages';

/**
 * Interface for validation result object
 * Requirement: Error Handling Matrix - Standardized validation responses
 */
export interface ValidationResult {
  isValid: boolean;
  errorCode?: string;
  errorMessage?: string;
}

/**
 * Interface for image validation options
 * Requirement: Data Protection Measures - File upload validation
 */
export interface ImageValidationOptions {
  maxSize: number;
  allowedTypes: string[];
  dimensions: {
    minWidth: number;
    minHeight: number;
    maxWidth?: number;
    maxHeight?: number;
  };
}

/**
 * Validates email format and checks for common email patterns
 * Requirements:
 * - Input validation and sanitization - Email validation
 * - Security Controls - Email format verification
 * @param email Email string to validate
 * @returns Boolean indicating if email is valid
 */
export const validateEmail = (email: string): boolean => {
  if (!email || typeof email !== 'string') {
    return false;
  }

  // Trim whitespace and convert to lowercase
  const normalizedEmail = email.trim().toLowerCase();

  // Check length constraints
  if (normalizedEmail.length < 5 || normalizedEmail.length > 254) {
    return false;
  }

  // Validate email format using validator library
  return validator.isEmail(normalizedEmail, {
    allow_utf8_local_part: false,
    require_tld: true,
    allow_ip_domain: false
  });
};

/**
 * Validates password strength and security requirements
 * Requirements:
 * - Input validation and sanitization - Password validation
 * - Security Controls - Password complexity verification
 * @param password Password string to validate
 * @returns Boolean indicating if password meets requirements
 */
export const validatePassword = (password: string): boolean => {
  if (!password || typeof password !== 'string') {
    return false;
  }

  // Check minimum length (8 characters)
  if (password.length < 8) {
    return false;
  }

  // Check for required character types
  const hasUppercase = /[A-Z]/.test(password);
  const hasLowercase = /[a-z]/.test(password);
  const hasNumber = /[0-9]/.test(password);
  const hasSpecial = /[!@#$%^&*(),.?":{}|<>]/.test(password);

  return hasUppercase && hasLowercase && hasNumber && hasSpecial;
};

/**
 * Validates phone number format and international standards
 * Requirements:
 * - Input validation and sanitization - Phone number validation
 * - Data Protection Measures - Phone format verification
 * @param phoneNumber Phone number string to validate
 * @returns Boolean indicating if phone number is valid
 */
export const validatePhoneNumber = (phoneNumber: string): boolean => {
  if (!phoneNumber || typeof phoneNumber !== 'string') {
    return false;
  }

  // Remove non-numeric characters
  const normalizedPhone = phoneNumber.replace(/\D/g, '');

  // Check length constraints (minimum 10, maximum 15 digits)
  if (normalizedPhone.length < 10 || normalizedPhone.length > 15) {
    return false;
  }

  // Validate using validator library's mobile phone check
  return validator.isMobilePhone(normalizedPhone, 'any', {
    strictMode: true
  });
};

/**
 * Validates geographic coordinates for furniture locations
 * Requirements:
 * - Input validation and sanitization - Coordinate validation
 * - Data Protection Measures - Location data validation
 * @param latitude Latitude coordinate
 * @param longitude Longitude coordinate
 * @returns Boolean indicating if coordinates are valid
 */
export const validateCoordinates = (latitude: number, longitude: number): boolean => {
  if (typeof latitude !== 'number' || typeof longitude !== 'number') {
    return false;
  }

  // Check latitude range (-90 to 90)
  if (latitude < -90 || latitude > 90) {
    return false;
  }

  // Check longitude range (-180 to 180)
  if (longitude < -180 || longitude > 180) {
    return false;
  }

  // Validate coordinate precision (maximum 6 decimal places)
  const latString = latitude.toString();
  const lonString = longitude.toString();
  const latDecimals = latString.includes('.') ? latString.split('.')[1].length : 0;
  const lonDecimals = lonString.includes('.') ? lonString.split('.')[1].length : 0;

  return latDecimals <= 6 && lonDecimals <= 6;
};

/**
 * Sanitizes user input to prevent XSS and injection attacks
 * Requirements:
 * - Input validation and sanitization - Input sanitization
 * - Security Controls - XSS prevention
 * @param input Input string to sanitize
 * @returns Sanitized input string
 */
export const sanitizeInput = (input: string): string => {
  if (!input || typeof input !== 'string') {
    return '';
  }

  // Trim whitespace
  let sanitized = input.trim();

  // Apply XSS filter
  sanitized = xss(sanitized, {
    whiteList: {}, // Disable all HTML tags
    stripIgnoreTag: true,
    stripIgnoreTagBody: ['script', 'style']
  });

  // Escape special characters
  sanitized = validator.escape(sanitized);

  return sanitized;
};

/**
 * Validates image file type, size, and dimensions
 * Requirements:
 * - Input validation and sanitization - File validation
 * - Data Protection Measures - Image upload validation
 * @param file File object to validate
 * @returns Boolean indicating if image file is valid
 */
export const validateImageFile = (file: any): boolean => {
  if (!file || typeof file !== 'object') {
    return false;
  }

  // Default validation options
  const options: ImageValidationOptions = {
    maxSize: 10 * 1024 * 1024, // 10MB
    allowedTypes: ['image/jpeg', 'image/png', 'image/webp'],
    dimensions: {
      minWidth: 200,
      minHeight: 200,
      maxWidth: 4096,
      maxHeight: 4096
    }
  };

  // Check file size
  if (file.size > options.maxSize) {
    return false;
  }

  // Validate mime type
  if (!options.allowedTypes.includes(file.type)) {
    return false;
  }

  // Check image dimensions (if available)
  if (file.width && file.height) {
    const { dimensions } = options;
    if (
      file.width < dimensions.minWidth ||
      file.height < dimensions.minHeight ||
      (dimensions.maxWidth && file.width > dimensions.maxWidth) ||
      (dimensions.maxHeight && file.height > dimensions.maxHeight)
    ) {
      return false;
    }
  }

  // Verify file integrity
  if (!file.buffer || !Buffer.isBuffer(file.buffer)) {
    return false;
  }

  return true;
};

File: src/android/app/src/main/kotlin/com/founditure/android/util/ValidationUtils.kt
/**
 * Human Tasks:
 * 1. Verify that the password requirements align with the security policy
 * 2. Ensure proper location permissions are configured in AndroidManifest.xml
 * 3. Review and adjust validation rules based on business requirements
 * 4. Configure proper error messages for different locales
 */

package com.founditure.android.util

import android.util.Patterns // version: latest
import android.location.Location // version: latest

/**
 * Utility class providing validation functions for user input, data models, and business rules.
 * Implements requirements from:
 * - Input Validation (7.3.3 Security Controls)
 * - Data Validation (3.6 Security Architecture)
 */
object ValidationUtils {

    /**
     * Validates email address format using Android patterns
     * Implements requirement: Input validation and sanitization for security
     */
    fun validateEmail(email: String): Boolean {
        if (email.isBlank()) {
            return false
        }
        return Patterns.EMAIL_ADDRESS.matcher(email).matches()
    }

    /**
     * Validates password strength according to security requirements
     * Implements requirement: Data validation for user inputs and model integrity
     */
    fun validatePassword(password: String): Boolean {
        if (password.length < 8) {
            return false
        }

        val hasUpperCase = password.any { it.isUpperCase() }
        val hasLowerCase = password.any { it.isLowerCase() }
        val hasDigit = password.any { it.isDigit() }
        val hasSpecialChar = password.any { !it.isLetterOrDigit() }

        return hasUpperCase && hasLowerCase && hasDigit && hasSpecialChar
    }

    /**
     * Validates phone number format using Android patterns
     * Implements requirement: Input validation and sanitization for security
     */
    fun validatePhoneNumber(phoneNumber: String): Boolean {
        if (phoneNumber.isBlank()) {
            return false
        }
        return Patterns.PHONE.matcher(phoneNumber).matches()
    }

    /**
     * Validates location coordinates and accuracy
     * Implements requirement: Data validation for user inputs and model integrity
     */
    fun validateLocation(latitude: Double, longitude: Double): Boolean {
        return latitude in -90.0..90.0 && longitude in -180.0..180.0
    }

    /**
     * Validates furniture listing input data
     * Implements requirements:
     * - Input validation and sanitization for security
     * - Data validation for user inputs and model integrity
     */
    fun validateFurnitureInput(
        title: String,
        description: String,
        images: List<String>,
        latitude: Double,
        longitude: Double
    ): Pair<Boolean, ErrorCodes?> {
        // Validate title
        if (title.isBlank() || title.length !in 3..100) {
            return Pair(false, ErrorCodes.VALIDATION_ERROR)
        }

        // Validate description
        if (description.isBlank() || description.length !in 10..1000) {
            return Pair(false, ErrorCodes.VALIDATION_ERROR)
        }

        // Validate images
        if (images.isEmpty()) {
            return Pair(false, ErrorCodes.VALIDATION_ERROR)
        }

        // Validate location
        if (!validateLocation(latitude, longitude)) {
            return Pair(false, ErrorCodes.LOCATION_ERROR)
        }

        return Pair(true, null)
    }

    // Private helper functions
    private fun containsOnlyAllowedCharacters(input: String): Boolean {
        val allowedPattern = Regex("^[a-zA-Z0-9\\s.,!?-]*$")
        return allowedPattern.matches(input)
    }

    private fun sanitizeInput(input: String): String {
        return input.trim()
            .replace(Regex("[<>\"'&]"), "")
            .take(1000) // Maximum length limit
    }
}